src/Core/Utils/BijectiveAssociation.hpp

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <initializer_list>
+#include <map>
+
+namespace Ra {
+namespace Core {
+namespace Utils {
+
+/**
+ * \brief Bijective association on a finite set of <key, value> pairs.
+ * \tparam T1 Type of the key
+ * \tparam T2 Type of the value
+ */
+template <typename T1, typename T2>
+class BijectiveAssociation
+{
+  public:
+    using key_type   = T1;
+    using value_type = T2;
+
+    /**
+     * \brief Constructor from { <T1, T2> } pairs
+     */
+    explicit BijectiveAssociation( std::initializer_list<std::pair<key_type, value_type>> pairs );
+
+    /**
+     * \brief Add a pair to the relation.
+     * \return false if the pair was not added due to collision with an already inserted pair
+     */
+    bool addAssociation( std::pair<key_type, value_type> p );
+
+    /**
+     * \brief Add a pair to the relation.
+     * \return false if the pair was not added due to collision with an already inserted pair
+     */
+    bool addAssociation( key_type v1, value_type v2 );
+
+    /**
+     * \brief Gets the value associated to the key
+     */
+    value_type operator()( const key_type& k ) const;
+
+    /**
+     * \brief Gets the value associated to the key
+     */
+    value_type value( const key_type& k ) const;
+
+    /**
+     * \brief Gets the key  associated to the value
+     */
+    key_type key( const value_type& k ) const;
+
+    /**
+     * \brief Gets an const iterator at beginning of the key to value map
+     */
+    typename std::map<key_type, value_type>::const_iterator begin();
```

Comment:
Well, I'll change this to `const_iterator begin() const noexcept;` (as in C++11 reference) to prevent the modification of the association through the iterator. The objective here is to have a read-only iterator.

Reply:
hence cbegin might be sufficient ?

Reply:
is cbegin sufficient for range based for ? I'll test it. But we can follow c++11 reference and define only 
```
const_iterator begin() const noexcept; 
const_iterator cbegin() const noexcept;
```

Reply:
Fine for me. And the user get a warning/error if it tries to get a non const iterator.

Reply:
`begin()` and `end()` functions are required for range-based for and algorithms. So, we have to define begin as well as cbegin with the above mentioned profiles.